libbeat/reader/multiline/counter.go

```diff
@@ -0,0 +1,131 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package multiline
+
+import (
+	"github.com/elastic/beats/v7/libbeat/reader"
+)
+
+type counterReader struct {
+	reader    reader.Reader
+	state     func(*counterReader) (reader.Message, error)
+	msgBuffer *messageBuffer
+}
+
+func newMultilineCountReader(
+	r reader.Reader,
+	separator string,
+	maxBytes int,
+	config *Config,
+) (reader.Reader, error) {
+	return &counterReader{
+		reader:    r,
+		state:     (*counterReader).readFirst,
+		msgBuffer: newMessageBuffer(maxBytes, config.LinesCount, []byte(separator), config.SkipNewLine),
```

Comment:
MaxLines vs. LinesCount:

The MaxLines and MaxBytes settings control truncation of the final event. LinesCount controls when to flush the message buffer and return an event. The multiline reader should not care about the state of the message buffer. Consider to introduce counters in the statemachine itself, to keep semantics of MaxBytes and MaxLines consistent with the pattern based multiline state machine.